servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/TimeoutHttpServiceFilter.java

```diff
@@ -0,0 +1,131 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.utils;
+
+import io.servicetalk.concurrent.api.Executor;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpExecutionStrategyInfluencer;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.http.utils.TimeoutHttpRequesterFilter.simpleDurationTimeout;
+
+/**
+ * A {@link StreamingHttpServiceFilter} that adds support for request/response timeouts.
+ *
+ * <p>The order with which this filter is applied may be highly significant. For example, appending it before a retry
+ * filter would have different results than applying it after the retry filter; timeout would apply for all retries vs
+ * timeout per retry.
+ */
+public final class TimeoutHttpServiceFilter
+        implements StreamingHttpServiceFilterFactory, HttpExecutionStrategyInfluencer {
+
+    private final TimeoutHttpRequesterFilter.TimeoutFromRequest timeoutForRequest;
+
+    @Nullable
+    private final Executor timeoutExecutor;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param duration the timeout {@link Duration}
+     */
+    public TimeoutHttpServiceFilter(Duration duration) {
+        this(simpleDurationTimeout(duration));
+    }
+
+    /**
+     * Construct a new instance.
+     *
+     * @param duration the timeout {@link Duration}
+     * @param timeoutExecutor the {@link Executor} to use for managing the timer notifications
+     */
+    public TimeoutHttpServiceFilter(Duration duration,
+                                    Executor timeoutExecutor) {
+        this(simpleDurationTimeout(duration), timeoutExecutor);
+    }
+
+    /**
+     * Construct a new instance.
+     *
+     * @param timeoutForRequest function for extracting timeout from request which may also determine the timeout using
+     * other sources. If no timeout is to be applied then the function should return null.
+     */
+    public TimeoutHttpServiceFilter(TimeoutHttpRequesterFilter.TimeoutFromRequest timeoutForRequest) {
```

Comment:
If you prefer a custom interface, consider making it a top level interface to avoid dependency between `TimeoutHttpServiceFilter` and `TimeoutHttpRequesterFilter`.